tools/migration/migrations/repo-1-2/migration.go

```diff
@@ -1 +1,380 @@
+// datastore. (taken from Store.writeTipSetAndState)
+func (m *MetadataFormatJSONtoCBOR) writeTipSetAndStateAsCBOR(tsas *chain.TipSetAndState) error {
+	if tsas.TipSetStateRoot == cid.Undef {
+		return errors.New("attempting to write state root cid.Undef")
+	}
+	val, err := cbor.DumpObject(tsas.TipSetStateRoot)
+	if err != nil {
+		return err
+	}
+
+	// datastore keeps tsKey:stateRoot (k,v) pairs.
+	h, err := tsas.TipSet.Height()
+	if err != nil {
+		return err
+	}
+	keyStr := makeKey(tsas.TipSet.String(), h)
+	key := datastore.NewKey(keyStr)
+
+	// this writes the value to the FSRepo
+	return m.store.Ds.Put(key, val)
+}
+
+// loadChainHeadAsJSON loads the latest known head from disk assuming JSON format
+func loadChainHeadAsJSON(chainStore *migrationChainStore) (types.SortedCidSet, error) {
+	return loadChainHead(false, chainStore)
+}
+
+// loadChainHeadAsCBOR loads the latest known head from disk assuming CBOR format
+func loadChainHeadAsCBOR(store *migrationChainStore) (types.SortedCidSet, error) {
+	return loadChainHead(true, store)
+}
+
+// loadChainHead loads the chain head CIDs as either CBOR or JSON
+func loadChainHead(asCBOR bool, store *migrationChainStore) (types.SortedCidSet, error) {
+	var emptyCidSet types.SortedCidSet
+
+	bb, err := store.Ds.Get(headKey)
+	if err != nil {
+		return emptyCidSet, errors.Wrap(err, "failed to read headKey")
+	}
+
+	var cids types.SortedCidSet
+	if asCBOR {
+		err = cbor.DecodeInto(bb, &cids)
+
+	} else {
+		err = json.Unmarshal(bb, &cids)
+	}
+	if err != nil {
+		return emptyCidSet, errors.Wrap(err, "failed to cast headCids")
+	}
+
+	return cids, nil
+}
+
+// function wrapper for loading state root as JSON
+func loadStateRootAsJSON(ts types.TipSet, store *migrationChainStore) (cid.Cid, error) {
+	return loadStateRoot(ts, false, store)
+}
+
+// loadStateRoot loads the chain store metadata into store, updating its
+// state root and then returning the state root + any error
+// pass true to load as CBOR (new format) or false to load as JSON (old format)
+func loadStateRoot(ts types.TipSet, asCBOR bool, store *migrationChainStore) (cid.Cid, error) {
+	h, err := ts.Height()
+	if err != nil {
+		return cid.Undef, err
+	}
+	key := datastore.NewKey(makeKey(ts.String(), h))
+	bb, err := store.Ds.Get(key)
+	if err != nil {
+		return cid.Undef, errors.Wrapf(err, "failed to read tipset key %s", ts.String())
+	}
+
+	var stateRoot cid.Cid
+	if asCBOR {
+		err = cbor.DecodeInto(bb, &stateRoot)
+	} else {
+		err = json.Unmarshal(bb, &stateRoot)
+	}
+	if err != nil {
+		return cid.Undef, errors.Wrapf(err, "failed to cast state root of tipset %s", ts.String())
+	}
+	return stateRoot, nil
+}
+
+// compareChainStores loads each chain store and iterates through each, comparing heads
+// then state root at each iteration, returning error if:
+//     * at any point state roots are not equal
+//     * the old and new chain store iterators are not complete at the same time
+// 	   * at the end the two heads are not equal
```

Comment:
Thanks, this is just the kind of comment I think is needed for an internal function like this.